pandas/core/arrays/categorical.py

```diff
@@ -671,6 +678,26 @@ def _codes_for_groupby(self, sort):
             categories in the original order.
         """
 
+        # we only care about observed values
+        if observed:
+            unique_codes = unique1d(self.codes)
```

Comment:
Haven't thought this through, but can this `if` block be replaced with `self.remove_unused_cateogories()._codes_for_groupby(sort=sort, observed=False)`?

Reply:
no, you actually need the uniques